core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -2036,7 +2037,7 @@ class KafkaController(val config: KafkaConfig,
   private def processFeatureUpdatesWithActiveController(request: UpdateFeaturesRequest,
                                                         callback: UpdateFeaturesCallback): Unit = {
     val updates = request.featureUpdates
-    val existingFeatures = featureCache.get
+    val existingFeatures = featureCache.getFeatureOption
```

Comment:
I realize this is existing code, but this seems incorrect. If someone changed the features in ZK recently, our cache is probably out of date. So we might lose an update by writing back the cache + whatever update is coming in.

I guess he is trying to avoid this with all the calls to `waitUntilFeatureEpochOrThrow` (bizarre pattern in my opinion...) but that doesn't really work because changes to the cache are async with regard to the ZK controller thread.

Reply:
I guess let's just file a follow-up JIRA and move on, given that we won't be using this any time soon.